Add support for SRv6 SID Manager

cscarpitta opened the pull request:
In the current FRR implementation, the SRv6 locator can only be used by one client at a time. This PR adds support for the SRv6 SID Manager, a zebra component responsible for SID allocation/management. The SRv6 SID Manager exposes a SID allocation/release APIs, allowing clients to request and release an SRv6 SID. The SID manager allows the SRv6 Locator to be shared among many clients such as IS-IS, BGP and others. The PR extends the BGP, IS-IS daemons to communicate with the SRv6 SID Manager to allocate/release SRv6 SIDs. 

The SID Manager supports two SRv6 SID allocation schemes:
-	_usid-f3216_: the SRv6 uSID allocation scheme. Ensures interoperability between FRR and other vendors including Cisco IOS XR and many others. The scheme has been used for the EANTC interoperability test (https://eantc.de/wp-content/uploads/2023/04/EANTC-InteropTest2023-TestReport.pdf). 
-	_uncompressed_: Ensures backward compatibility with the current implementation of FRR.

ton31337 commented on lib/zclient.c:
```diff
@@ -3264,6 +3274,270 @@ int srv6_manager_release_locator_chunk(struct zclient *zclient,
+		flog_err(EC_LIB_ZAPI_ENCODE,
+			 "Invalid SRv6 SID: %pI6 ctx %s, SID context mismatch %s != %s",
+			 &response_sid_value,
+			 srv6_sid_ctx2str(buf, sizeof(buf), &response_ctx),
```
Seems this is a duplicate output, maybe we can have it printed once?

ton31337 commented on lib/zclient.c:
```diff
@@ -3264,6 +3274,270 @@ int srv6_manager_release_locator_chunk(struct zclient *zclient,
+			   sizeof(struct in6_addr)) != 0) {
+			flog_err(EC_LIB_ZAPI_ENCODE,
+				 "Invalid SRv6 SID: %pI6 ctx %s, SID value mismatch %pI6 != %pI6",
+				 &response_sid_value,
```
Same here.

ton31337 commented on lib/zclient.c:
```diff
@@ -3264,6 +3274,270 @@ int srv6_manager_release_locator_chunk(struct zclient *zclient,
+		return -1;
+	}
+	if (!sid_zero_ipv6(sid_value)) {
+		if (memcmp(sid_value, &response_sid_value,
```
Can we use IPV6_ADDR_SAME()?

cscarpitta replied:
Done!

ton31337 reviewed (changes requested):
Documentation is missing yet. Another question: is this similar to the label allocation manager (synchronous)?

dmytroshytyi-6WIND commented on lib/zclient.c:
```diff
@@ -3264,6 +3274,270 @@ int srv6_manager_release_locator_chunk(struct zclient *zclient,
+	/* Put length at the first point of the stream. */
+	stream_putw_at(s, 0, stream_get_endp(s));
+
+	return zclient_send_message(zclient);
```
This is comment [0]
In this place, behind this func,
`event_add_write()` - it seems, like this func is called.

Please see the comment [1] below that is in relation with current commit.

cscarpitta replied:
Initially, I designed this API to be synchronous.
Now, based on the feedback of the FRR community, I updated the API to be synchronous.

So, actually `zclient_read_sync_response` is no longer used.

dmytroshytyi-6WIND commented:
How do we set a preference between compressed and uncompressed SRv6 SID selection?  What is the default method?

dmytroshytyi-6WIND commented:
How the unsetting the format is expected to work, do you verify unsetting of f3216 in topotest?

dmytroshytyi-6WIND commented:
What are the requirements to successfully run the [topotest ](https://github.com/FRRouting/frr/pull/15679)that verifies the functionality of this PR?

dmytroshytyi-6WIND commented:
This is more like an open question: it might be interesting to create a new PR (1 and complete) that combines 4 PRs related to the SID manager as a draft? What stops us to do that?

cscarpitta commented on zebra/zebra_srv6.c:
```diff
@@ -204,7 +232,7 @@ void zebra_notify_srv6_locator_delete(struct srv6_locator *locator)
 	 * owner of each chunk.
 	 */
-	for (ALL_LIST_ELEMENTS_RO((struct list *)locator->chunks, n, c)) {
+	for (ALL_LIST_ELEMENTS_RO((struct list *)locator->locator.chunks, n, c)) {
```
Agree.

I fixed this by adding the required fields to the existing `srv6_locator` data structure instead of creating a nested data structure.

Looks better now.

cscarpitta commented on lib/srv6.h:
```diff
@@ -246,6 +257,52 @@ const char *seg6local_context2str(char *str, size_t size,
+	case ZEBRA_SEG6_LOCAL_ACTION_END_AS:
+	case ZEBRA_SEG6_LOCAL_ACTION_END_AM:
+	case ZEBRA_SEG6_LOCAL_ACTION_END_BPF:
+	case ZEBRA_SEG6_LOCAL_ACTION_UNSPEC:
```
Correct, it's a bug. Fixed!

cscarpitta commented on lib/zclient.c:
```diff
@@ -1124,6 +1124,11 @@ int zapi_srv6_locator_encode(struct stream *s, const struct srv6_locator *l)
+	stream_putc(s, l->function_bits_length);
+	stream_putc(s, l->argument_bits_length);
+	stream_putc(s, l->flags);
 	return 0;
```
zebra uses these encode/decode functions to communicate to the locator information to protocols (BGP, ISIS, ...).

zebra needs to provide also block/node/func lengths, because protocols must advertise them in the corresponding TLVs

That's the reason why we need to include block/node/func lengths in the encode/decode functions.

cscarpitta commented on lib/zclient.c:
```diff
@@ -3264,6 +3274,270 @@ int srv6_manager_release_locator_chunk(struct zclient *zclient,
+ * @param sid_func SID function assigned by the SRv6 Manager
+ * @result 0 on success, -1 otherwise
+ */
+int srv6_manager_get_srv6_sid(struct zclient *zclient,
```
Fair point. I can shorten the function names in this way:

`srv6_manager_get_srv6_sid -> srv6_manager_get_sid`

cscarpitta commented on lib/zclient.c:
```diff
@@ -3264,6 +3274,270 @@ int srv6_manager_release_locator_chunk(struct zclient *zclient,
+ * @param ctx Context associated with the SRv6 SID to be removed
+ * @result 0 on success, -1 otherwise
+ */
+int sm_release_srv6_sid(struct zclient *zclient, const struct srv6_sid_ctx *ctx)
```
Yes. The two functions are correlated

`srv6_manager_get_srv6_sid` requests a SID to SID Manager
`sm_release_srv6_sid` release a SID to SID Manager

I will rename these functions as follows:
`srv6_manager_get_srv6_sid -> srv6_manager_get_sid`
`sm_release_srv6_sid -> srv6_manager_release_sid`

This way it is clear that the two functions are correlated

cscarpitta commented on lib/zclient.h:
```diff
@@ -1070,10 +1073,22 @@ extern int tm_get_table_chunk(struct zclient *zclient, uint32_t chunk_size,
+			      const struct srv6_sid_ctx *ctx,
+			      struct in6_addr *sid_value,
+			      const char *locator_name, uint32_t *sid_func);
+int sm_release_srv6_sid(struct zclient *zclient, const struct srv6_sid_ctx *ctx);
```
Changed to `srv6_manager_release_sid`, which looks more clear

cscarpitta commented on lib/zclient.c:
```diff
@@ -3264,6 +3274,270 @@ int srv6_manager_release_locator_chunk(struct zclient *zclient,
+		zlog_debug("SRv6 SID request (%d bytes) sent", ret);
+
+	/* read response */
+	if (zclient_read_sync_response(zclient,
```
Initially, I designed this API to be synchronous.
Now, based on the feedback of the FRR community, I updated the API to be synchronous.

So, actually `zclient_read_sync_response` is no longer used.

cscarpitta commented on lib/srv6.h:
```diff
@@ -246,6 +257,52 @@ const char *seg6local_context2str(char *str, size_t size,
+	case ZEBRA_SEG6_LOCAL_ACTION_END_AS:
+	case ZEBRA_SEG6_LOCAL_ACTION_END_AM:
+	case ZEBRA_SEG6_LOCAL_ACTION_END_BPF:
+	case ZEBRA_SEG6_LOCAL_ACTION_UNSPEC:
```
These behaviors are treated as unknown because they are currently not supported in FRR.

cscarpitta commented:
> Documentation is missing yet. Another question: is this similar to the label allocation manager (synchronous)?

Hi @ton31337 many thanks for the review. I added the documentation.

Regarding the second question, initially it was similar to the label allocation manager (synchronous)
I received some feedbacks from the community, it was suggested to me to make it asynchronous.

I updated the PR based on the feedback. Now it is asynchronous.

pguibert6WIND commented:
there are some address sanitizer error : https://ci1.netdef.org/browse/FRR-PULLREQ3-3112/artifact/ASAN7/AddressSanitizerError/AddressSanitzer.txt , and it seems related to your dev.
please use `./configure ... --enable-address-sanitizer` and run the topotest to see asan files in /tmp/topotests/<yourtest>/.

cscarpitta commented:
@pguibert6WIND Thanks for the review. I fixed the errors reported by address sanitizer.

riw777 reviewed (commented):
I think we'll need topo tests for this one (?) ... other than that, the code looks fine

cscarpitta commented:
> I think we'll need topo tests for this one (?) ... other than that, the code looks fine

@riw777 Many thanks for the review.

We have another PR with the topotest (https://github.com/FRRouting/frr/pull/15679). The topotest extensively tests the SID manager and the interaction with the client daemons (BGP, IS-IS).

Please note, originally the SID manager, client daemon extension (BGP, IS-IS) and the topotest were in a single PR.
But the PR was very large, hence we split into SID Manager, BGP extension, IS-IS extension, and topotest.

riw777 reviewed (approved):
looks good

dmytroshytyi-6WIND reviewed (approved):
LGTM
